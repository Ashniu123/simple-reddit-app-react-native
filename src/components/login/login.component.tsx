import * as React from 'react';
import { WebView, Alert } from 'react-native';
import { connect } from 'react-redux';

import styles from './login.style';

import { startLoginUser, startLoginIfTokenExists } from '../../actions/auth';

export class Login extends React.Component {
	static navigationOptions = {
		title: 'Login to Reddit',
		headerTitleStyle: styles.header__text
	};

	constructor(props) {
		super(props);

		this.handleNavigationStateChange = this.handleNavigationStateChange.bind(this);

		if(this.props.token && this.props.token.length > 0) {
			this.props.navigation.navigate('App');
		}
	}

	async componentDidMount() {
		try {
			await this.props.startLoginIfTokenExists();
		} catch(err) {
			Alert.alert(JSON.stringify(err));
		}
	}

	handleNavigationStateChange(navState) {
		if (navState.url.indexOf('callback://login#') === 0) {
			const accessToken: string = navState.url.match(/^callback:\/\/login#access_token=(.+)&token.*/)[1];
			this.props.startLoginUser(accessToken)
				.then(() => {
					this.props.navigation.navigate('App');
				})
				.catch((err) => {
					Alert.alert('Invalid auth');
					console.log(err);
				});
		}
	}

	render() {
		const REDDIT_APP_ID: string = 'V99jnyVaW3mbNQ';
		const LOGIN_URL: string = `https://www.reddit.com/api/v1/authorize?client_id=${REDDIT_APP_ID}&response_type=token&
		state=RANDOM_STRING&redirect_uri=callback://login&scope=read`;
		return <WebView
			source={{ uri: LOGIN_URL }}
			onNavigationStateChange={this.handleNavigationStateChange}
		/>
	}
}

const mapStateToProps = (state) => ({
	token: state.auth.token
});

const mapDispatchToProps = (dispatch) => ({
	startLoginUser: (token) => dispatch(startLoginUser(token)),
	startLoginIfTokenExists: () => dispatch(startLoginIfTokenExists())
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
