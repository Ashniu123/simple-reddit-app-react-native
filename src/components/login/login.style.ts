import { StyleSheet } from 'react-native';

export default StyleSheet.create({
	header__text: {
		textAlign: 'center'
	}
});