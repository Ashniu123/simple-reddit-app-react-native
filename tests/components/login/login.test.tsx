import * as React from 'react';
import { shallow } from 'enzyme';

import { Login } from '../../../src/components/login/login.component';

let startLoginUser, startLoginIfTokenExists, navigation;

beforeEach(() => {
	startLoginUser = jest.fn();
	startLoginIfTokenExists = jest.fn();
	navigation = { navigate: jest.fn() }
});

test('should render Login component correctly', () => {
	const wrapper = shallow(<Login
		startLoginUser={startLoginUser}
		startLoginIfTokenExists={startLoginIfTokenExists}
		navigation={navigation}
	/>);
	expect(startLoginIfTokenExists).toHaveBeenCalled();
	expect(wrapper).toMatchSnapshot();
});

test('should redirect to "App" if token exists', () => {
	const token = 'Some Token';
	const wrapper = shallow(<Login
		startLoginUser={startLoginUser}
		startLoginIfTokenExists={startLoginIfTokenExists}
		navigation={navigation}
		token={token}
	/>);
	expect(navigation.navigate).toHaveBeenCalledWith('App');
});

test('should not redirect to "App" if token not exists', () => {
	const token = '';
	const wrapper = shallow(<Login
		startLoginUser={startLoginUser}
		startLoginIfTokenExists={startLoginIfTokenExists}
		navigation={navigation}
		token={token}
	/>);
	expect(navigation.navigate).not.toHaveBeenCalled();
});