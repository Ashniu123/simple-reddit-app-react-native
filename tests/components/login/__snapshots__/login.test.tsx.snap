// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login component correctly 1`] = `
<WebView
  onNavigationStateChange={[Function]}
  scalesPageToFit={true}
  source={
    Object {
      "uri": "https://www.reddit.com/api/v1/authorize?client_id=V99jnyVaW3mbNQ&response_type=token&
		state=RANDOM_STRING&redirect_uri=callback://login&scope=read",
    }
  }
/>
`;
